utils/global-state-update-gen/src/generic.rs

```diff
@@ -237,15 +237,20 @@ pub fn add_and_remove_bids<T: StateReader>(
     };
 
     for (pub_key, seigniorage_recipient) in new_snapshot.values().next().unwrap() {
-        let stake = *seigniorage_recipient.stake();
-        create_or_update_bid(state, pub_key, stake);
+        create_or_update_bid(state, pub_key, seigniorage_recipient);
     }
 
     // Refresh the bids - we modified them above.
     let bids = state.get_bids();
 
     for pub_key in to_unbid {
         if let Some(bid) = bids.get(&pub_key) {
+            for delegator in bid.delegators().values() {
+                // Burn the delegated funds of all the delegators.
+                // TBD: is that what should be happening when a validator is removed?
```

Comment:
IMO funds should be returned to the delegators but we need to decide before merging this.

Reply:
I'm actually leaning towards burning. That's because this is mostly intended to be used in cases where we _have_ to remove a validator, because it was misbehaving and was part of a problem that brought the network down. If we didn't burn delegators' stakes, the validator could defend against losing their stake after misbehaving by only staking a small amount themselves, and delegating almost all of the stake from a different account.

Reply:
Because otherwise it leaves an easy way of protecting the stake for validators that are planning to misbehave.
As in - if you stake 1000000 CSPR and misbehave, you lose 1000000 CSPR. But if you stake 100 CSPR, then delegate the remaining 999900 CSPR from a different account, your total stake in the network is the same, but if you misbehave, you only lose 100 CSPR.
But I'm not sure if that's a sufficient argument for doing that.

Reply:
I understand how the malicious validators could avoid being slashed but I'm still not convinced that the cure isn't worse than disease.

Reply:
Yeah, I'm not sure myself. I guess it could be seen as incentivizing only delegating to validators you trust, and also to validators that have high stakes themselves.